File: src/main/java/ru/citeck/ecos/settings/EcosServerColumnInfo.java
package ru.citeck.ecos.settings;

import com.intellij.openapi.util.NlsContexts;
import com.intellij.ui.components.JBTextField;
import java.util.function.Function;
import com.intellij.util.ui.ColumnInfo;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.table.TableCellEditor;
import java.util.function.BiConsumer;

final class EcosServerColumnInfo extends ColumnInfo<EcosServer, String> {

    private final Function<EcosServer, String> getter;
    private final BiConsumer<EcosServer, String> setter;

    private TableCellEditor editor = new DefaultCellEditor(new JBTextField());
    private Function<EcosServer, Boolean> cellEditableEvaluator = ecosServer -> true;

    public EcosServerColumnInfo(
            @NlsContexts.ColumnName String name,
            @NotNull Function<EcosServer, String> getter,
            BiConsumer<EcosServer, String> setter
    ) {
        super(name);
        this.getter = getter;
        this.setter = setter;
    }

    public EcosServerColumnInfo withEditor(TableCellEditor editor) {
        this.editor = editor;
        return this;
    }

    public EcosServerColumnInfo withCellEditableEvaluator(Function<EcosServer, Boolean> cellEditableEvaluator) {
        this.cellEditableEvaluator = cellEditableEvaluator;
        return this;
    }

    @Override
    public @NotNull TableCellEditor getEditor(EcosServer ecosServer) {
        return editor;
    }

    @Override
    public @Nullable String valueOf(EcosServer ecosServer) {
        return getter.apply(ecosServer);
    }

    @Override
    public boolean isCellEditable(EcosServer ecosServer) {
        return cellEditableEvaluator.apply(ecosServer);
    }

    @Override
    public void setValue(EcosServer ecosServer, String value) {
        setter.accept(ecosServer, value);
    }

}

File: src/main/kotlin/ru/citeck/idea/settings/servers/EcosServerColumnInfo.kt
package ru.citeck.idea.settings.servers

import com.intellij.openapi.util.NlsContexts.ColumnName
import com.intellij.ui.components.JBTextField
import com.intellij.util.ui.ColumnInfo
import javax.swing.DefaultCellEditor
import javax.swing.table.DefaultTableCellRenderer
import javax.swing.table.TableCellEditor
import javax.swing.table.TableCellRenderer

internal class EcosServerColumnInfo(
    name: @ColumnName String,
    private val getter: (CiteckServer.Builder) -> String,
    private val setter: (CiteckServer.Builder, String) -> Unit
) : ColumnInfo<CiteckServer.Builder, String>(name) {

    private var editor: TableCellEditor = DefaultCellEditor(JBTextField())
    private var renderer: TableCellRenderer = DefaultTableCellRenderer()

    fun withEditor(editor: TableCellEditor): EcosServerColumnInfo {
        this.editor = editor
        return this
    }

    fun withRenderer(renderer: TableCellRenderer): EcosServerColumnInfo {
        this.renderer = renderer
        return this
    }

    override fun getEditor(citeckServer: CiteckServer.Builder): TableCellEditor {
        return editor
    }

    override fun valueOf(citeckServer: CiteckServer.Builder): String {
        return getter.invoke(citeckServer)
    }

    override fun isCellEditable(citeckServer: CiteckServer.Builder): Boolean {
        return true
    }

    override fun setValue(citeckServer: CiteckServer.Builder, value: String?) {
        setter.invoke(citeckServer, value ?: "")
    }

    override fun getRenderer(citeckServer: CiteckServer.Builder): TableCellRenderer {
        return renderer
    }
}
